File: bfs/icebreg.cpp
#include<bits/stdc++.h>
using namespace std;

int board[302][302];
int vis[302][302];
int dx[4] = {1,0,-1,0};
int dy[4] = {0,1,0,-1};

int bfs(int n, int m) {
    int breg = 0;
    for(int i=0;i<n;i++) {
        for(int j=0;j<m;j++) {
            vis[i][j] = 0;
        }
    }
    for(int i=0;i<n;i++) {
        for(int j=0;j<m;j++) {
            if(board[i][j]>0&&vis[i][j]!=1) {
                vis[i][j] = 1;
                breg++;
                queue<pair<int, int>> Q;
                Q.push({i,j});
                while(!Q.empty()) {
                    pair<int, int> cur = Q.front();
                    Q.pop();
                    for(int dir=0;dir<4;dir++) {
                        int nx = cur.first + dx[dir];
                        int ny = cur.second + dy[dir];
                        if(nx<0||nx>=n||ny<0||ny>=m) continue;
                        if(vis[nx][ny]==1||board[nx][ny]<=0) continue;
                        vis[nx][ny] = 1;
                        Q.push({nx, ny});
                    }
                }
            }
        }
    }
    return breg;
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);

    int n,m;
    cin>>n>>m;
    for(int i=0;i<n;i++) {
        for(int j=0;j<m;j++) {
            cin>>board[i][j];
        }
    }
    int year=0;
    while(true) {
        int regions = bfs(n,m);
        if(regions>1) {
            printf("%d", year);
            break;
        } else if(regions==0) {
            printf("0");
            break;
        }
        year++;

        // 임시 배열을 사용해서 동시에 녹이기
        int temp[302][302];
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                temp[i][j] = board[i][j];
            }
        }

        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                if(board[i][j]>0){
                    int count=0;
                    for(int dir=0;dir<4;dir++) {
                        int nx = i + dx[dir];
                        int ny = j + dy[dir];
                        if(nx<0||nx>=n||ny<0||ny>=m) continue;
                        if(board[nx][ny]<=0) count++;
                    }
                    temp[i][j] = max(0, board[i][j] - count);
                }
            }
        }
        
        // temp 배열을 board로 복사
        for(int i=0;i<n;i++) {
            for(int j=0;j<m;j++) {
                board[i][j] = temp[i][j];
            }
        }
    }
}

File: bfs/icebreg.py
import sys
input = sys.stdin.readline
from collections import deque

def bfs(sea, n, m):
    vis = [[0]*m for _ in range(n)]
    dx = [-1, 0, 1, 0]
    dy = [0, -1, 0, 1]
    
    queue = deque()
    count = 0
    
    for i in range(n):
        for j in range(m):
            if sea[i][j] != 0 and vis[i][j] == 0:
                count += 1
                queue.append((i, j))
                vis[i][j] = 1
                
                while queue:
                    x, y = queue.popleft()
                    for dir in range(4):
                        nx = x + dx[dir]
                        ny = y + dy[dir]
                        if nx < 0 or nx >= n or ny < 0 or ny >= m:
                            continue
                        if sea[nx][ny] != 0 and vis[nx][ny] == 0:
                            queue.append((nx, ny))
                            vis[nx][ny] = 1
    
    return count

def main():
    n, m = map(int, input().split())
    sea = [list(map(int, input().split())) for _ in range(n)]
    dx = [-1, 0, 1, 0]
    dy = [0, -1, 0, 1]
    
    year = 0
    
    while True:
        ice_count = bfs(sea, n, m)
        
        if ice_count >= 2:
            print(year)
            break
        elif ice_count == 0:
            print(0)
            break

        new_sea = [row[:] for row in sea]
        
        for i in range(n):
            for j in range(m):
                if sea[i][j] != 0:
                    melting = 0
                    for dir in range(4):
                        nx = i + dx[dir]
                        ny = j + dy[dir]
                        if nx < 0 or nx >= n or ny < 0 or ny >= m:
                            continue
                        if sea[nx][ny] == 0:
                            melting += 1
                    new_sea[i][j] = max(0, sea[i][j] - melting)
        
        sea = new_sea
        year += 1

if __name__ == "__main__":
    main()
